gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -157,32 +164,111 @@ public void handleResponse(WriteObjectResponse response) {
+    if (!isOpen()) {
+      throw new ClosedChannelException();
+    }
+
+    // No point in writing further if upload failed on another thread.
+    if (uploadOperation.isDone()) {
+      waitForCompletionAndThrowIfUploadFailed();
+    }
+
+    while (buffer.remaining() > 0) {
+      int bytesToWrite = 0;
+      if (lastChunk != null) {
+        bytesToWrite = min(buffer.remaining(), MAX_BYTES_PER_MESSAGE - lastChunk.size());
+        lastChunk = lastChunk.concat(ByteString.copyFrom(buffer, bytesToWrite));
+      } else {
+        bytesToWrite = min(buffer.remaining(), MAX_BYTES_PER_MESSAGE);
+        lastChunk = ByteString.copyFrom(buffer, bytesToWrite);
+      }
+      bytesWritten += bytesToWrite;
+      if (lastChunk.size() == MAX_BYTES_PER_MESSAGE) {
+        try {
+          writeBuffer.put(lastChunk);
+          lastChunk = null;
+        } catch (InterruptedException e) {
+          throw new IOException(
+              String.format(
+                  "Failed to write %d bytes in '%s'", buffer.remaining(), getResourceString()),
+              e);
+        }
+      }
+    }
+
+    if (uploadOperation.isDone()) {
+      waitForCompletionAndThrowIfUploadFailed();
+    }
+    return bytesWritten;
+  }
+
+  /** Initialize this channel object for writing. */
+  public void initialize() throws IOException {
+    writeBuffer =
+        new LinkedBlockingQueue<ByteString>(
+            Math.toIntExact(channelOptions.getNumberOfBufferedRequests()));
     try {
-      uploadOperation = threadPool.submit(new UploadOperation(pipeSource));
+      uploadOperation = threadPool.submit(new UploadOperation(writeBuffer));
     } catch (Exception e) {
       throw new RuntimeException(String.format("Failed to start upload for '%s'", resourceId), e);
     }
+    initialized = true;
+  }
+
+  public void startUpload(InputStream pipeSource) throws IOException {
+    return;
+  }
+
+  @Override
+  public void close() throws IOException {
+    checkState(initialized, "initialize() must be invoked before use.");
+    if (!isOpen()) {
+      return;
+    }
+    if (lastChunk == null) {
+      lastChunk = ByteString.EMPTY;
+    }
+    try {
+      writeBuffer.put(lastChunk);
+      lastChunk = null;
+      handleResponse(waitForCompletionAndThrowIfUploadFailed());
+    } catch (IOException e) {
+      throw e;
+    } catch (InterruptedException e) {
+      // FIXME: handle the exception
+      e.printStackTrace();
+    } finally {
+      writeBuffer = null;
+      closeInternal();
+    }
   }
 
   private class UploadOperation implements Callable<WriteObjectResponse> {
 
     // Read end of the pipe.
-    private final BufferedInputStream pipeSource;
     private final int MAX_BYTES_PER_MESSAGE = MAX_WRITE_CHUNK_BYTES.getNumber();
 
     private Hasher objectHasher;
     private String uploadId;
     private long writeOffset = 0;
-    private InsertChunkResponseObserver responseObserver;
+    private LinkedBlockingQueue<ByteString> writeBuffer;
```

Comment:
It is bit confusing to have a variable with the same name in the inner class. Consider renaming.